File: go/data/builtins.go
package data

import (
	"errors"
	"fmt"
)

func print(args []Value) (Value, error) {
	if len(args) != 1 {
		return nil, fmt.Errorf("print got %d arg(s); want 1", len(args))
	}
	fmt.Println(args[0])
	return TheUndef, nil
}

func parseInts(name string, args []Value) ([]int, error) {
	if len(args) == 0 {
		return nil, fmt.Errorf("%s got %d arg(s); want 1+", name, len(args))
	}
	var is []int
	for _, arg := range args {
		v, ok := arg.(*Int)
		if !ok {
			return nil, fmt.Errorf("%s failed: not an int", name)
		}
		is = append(is, v.I)
	}
	return is, nil
}

func add(args []Value) (Value, error) {
	is, err := parseInts("+", args)
	if err != nil {
		return nil, err
	}
	a := 0
	for _, i := range is {
		a += i
	}
	return NewInt(a), nil
}

func sub(args []Value) (Value, error) {
	is, err := parseInts("-", args)
	if err != nil {
		return nil, err
	}
	a := is[0]
	for _, i := range is[1:] {
		a -= i
	}
	return NewInt(a), nil
}

func mul(args []Value) (Value, error) {
	is, err := parseInts("*", args)
	if err != nil {
		return nil, err
	}
	a := 1
	for _, i := range is {
		a *= i
	}
	return NewInt(a), nil
}

func div(args []Value) (Value, error) {
	is, err := parseInts("/", args)
	if err != nil {
		return nil, err
	}
	a := is[0]
	for _, i := range is[1:] {
		a /= i
	}
	return NewInt(a), nil
}

func parseTwoInts(name string, args []Value) (int, int, error) {
	if len(args) != 2 {
		return 0, 0, fmt.Errorf("%s got %d arg(s); want 2", name, len(args))
	}
	lhs, ok := args[0].(*Int)
	if !ok {
		return 0, 0, fmt.Errorf("%s failed: not an int", name)
	}
	rhs, ok := args[1].(*Int)
	if !ok {
		return 0, 0, fmt.Errorf("%s failed: not an int", name)
	}
	return lhs.I, rhs.I, nil
}

func eq(args []Value) (Value, error) {
	lhs, rhs, err := parseTwoInts("=", args)
	if err != nil {
		return nil, err
	}
	return NewBool(lhs == rhs), nil
}

func lt(args []Value) (Value, error) {
	lhs, rhs, err := parseTwoInts("<", args)
	if err != nil {
		return nil, err
	}
	return NewBool(lhs < rhs), nil
}

func lte(args []Value) (Value, error) {
	lhs, rhs, err := parseTwoInts("<=", args)
	if err != nil {
		return nil, err
	}
	return NewBool(lhs <= rhs), nil
}

func gt(args []Value) (Value, error) {
	lhs, rhs, err := parseTwoInts(">", args)
	if err != nil {
		return nil, err
	}
	return NewBool(lhs > rhs), nil
}

func gte(args []Value) (Value, error) {
	lhs, rhs, err := parseTwoInts(">=", args)
	if err != nil {
		return nil, err
	}
	return NewBool(lhs >= rhs), nil
}

func and(args []Value) (Value, error) {
	result := true
	for _, arg := range args {
		result = result && arg != TheFalse
	}
	return NewBool(result), nil
}

func or(args []Value) (Value, error) {
	result := false
	for _, arg := range args {
		result = result || arg != TheFalse
	}
	return NewBool(result), nil
}

func not(args []Value) (Value, error) {
	if len(args) != 1 {
		return nil, fmt.Errorf("not got %d arg(s); want 1", len(args))
	}
	return NewBool(!(args[0] != TheFalse)), nil
}

func eqCheck(args []Value) (Value, error) {
	if len(args) != 2 {
		return nil, fmt.Errorf("eq? got %d arg(s); want 2", len(args))
	}
	return NewBool(args[0] == args[1]), nil
}

func cons(args []Value) (Value, error) {
	if len(args) != 2 {
		return nil, fmt.Errorf("cons got %d arg(s); want 2", len(args))
	}
	return NewPair(args[0], args[1]), nil
}

func car(args []Value) (Value, error) {
	if len(args) != 1 {
		return nil, fmt.Errorf("car got %d arg(s); want 1", len(args))
	}
	pair, ok := args[0].(*Pair)
	if !ok {
		return nil, errors.New("car failed: not a pair")
	}
	return pair.Car, nil
}

func cdr(args []Value) (Value, error) {
	if len(args) != 1 {
		return nil, fmt.Errorf("cdr got %d arg(s); want 1", len(args))
	}
	pair, ok := args[0].(*Pair)
	if !ok {
		return nil, errors.New("cdr failed: not a pair")
	}
	return pair.Cdr, nil
}

type nativeFunc struct {
	name string
	fun  func(args []Value) (Value, error)
}

func (f *nativeFunc) String() string {
	return f.name
}

func (f *nativeFunc) Apply(args []Value) (Value, error) {
	return f.fun(args)
}

var builtins = map[string]FuncValue{
	"print": &nativeFunc{"print", print},
	"+":     &nativeFunc{"+", add},
	"-":     &nativeFunc{"-", sub},
	"*":     &nativeFunc{"*", mul},
	"/":     &nativeFunc{"/", div},
	"=":     &nativeFunc{"=", eq},
	"<":     &nativeFunc{"<", lt},
	"<=":    &nativeFunc{"<=", lte},
	">":     &nativeFunc{">", gt},
	">=":    &nativeFunc{">=", gte},
	"and":   &nativeFunc{"and", and},
	"or":    &nativeFunc{"or", or},
	"not":   &nativeFunc{"not", not},
	"eq?":   &nativeFunc{"eq?", eqCheck},
	"cons":  &nativeFunc{"cons", cons},
	"car":   &nativeFunc{"car", car},
	"cdr":   &nativeFunc{"cdr", cdr},
}

File: javascript/src/builtins.js
const data = require('./data.js');

function print(value) {
  console.log(value.toString());
  return data.theUndef;
}

function unwrapInt(value) {
  if (!(value instanceof data.Int)) {
    throw new Error('not an int');
  }
  return value.rawValue;
}

function add(...args) {
  if (args.length === 0) {
    throw new Error('+ got 0 arg');
  }
  let a = 0;
  for (const arg of args) {
    a += unwrapInt(arg);
  }
  return new data.Int(a);
}

function sub(...args) {
  if (args.length === 0) {
    throw new Error('- got 0 arg');
  }
  let a = unwrapInt(args[0]);
  for (const arg of args.slice(1)) {
    a -= unwrapInt(arg);
  }
  return new data.Int(a);
}

function mul(...args) {
  if (args.length === 0) {
    throw new Error('* got 0 arg');
  }
  let a = 1;
  for (const arg of args) {
    a *= unwrapInt(arg);
  }
  return new data.Int(a);
}

function div(...args) {
  if (args.length === 0) {
    throw new Error('/ got 0 arg');
  }
  let a = unwrapInt(args[0]);
  for (const arg of args.slice(1)) {
    a /= unwrapInt(arg);
  }
  return new data.Int(a);
}

function eq(a, b) {
  return data.createBool(unwrapInt(a) === unwrapInt(b));
}

function lt(a, b) {
  return data.createBool(unwrapInt(a) < unwrapInt(b));
}

function lte(a, b) {
  return data.createBool(unwrapInt(a) <= unwrapInt(b));
}

function gt(a, b) {
  return data.createBool(unwrapInt(a) > unwrapInt(b));
}

function gte(a, b) {
  return data.createBool(unwrapInt(a) >= unwrapInt(b));
}

function and(...args) {
  let result = true;
  for (const arg of args) {
    result = result && !arg.equals(data.theFalse);
  }
  return data.createBool(result);
}

function or(...args) {
  let result = false;
  for (const arg of args) {
    result = result || !arg.equals(data.theFalse);
  }
  return data.createBool(result);
}

function not(value) {
  return data.createBool(value.equals(data.theFalse));
}

function eqCheck(a, b) {
  return data.createBool(a.equals(b));
}

function cons(car, cdr) {
  return new data.Pair(car, cdr);
}

function car(value) {
  if (!(value instanceof data.Pair)) {
    throw new Error('not a pair');
  }
  return value.car;
}

function cdr(value) {
  if (!(value instanceof data.Pair)) {
    throw new Error('not a pair');
  }
  return value.cdr;
}

const ALL = new Map();

function register(name, func) {
  ALL.set(name, new data.Func(name, func));
}

register('print', print);
register('+', add);
register('-', sub);
register('*', mul);
register('/', div);
register('=', eq);
register('<', lt);
register('<=', lte);
register('>', gt);
register('>=', gte);
register('and', and);
register('or', or);
register('not', not);
register('eq?', eqCheck);
register('cons', cons);
register('car', car);
register('cdr', cdr);

Object.assign(module.exports, {
  ALL,
});
